rfcs/20200721-extension-types.md

```diff
@@ -0,0 +1,1304 @@
+`tf.SavedModel`) and lower-level APIs (such as `tf.while_loop` and `tf.add`).  
+
+This RFC proposes a set of protocols that will allow TensorFlow APIs to handle
+user-defined Python types.  A version of this interface is already used
+internally to implement several core TensorFlow data types, including
+`tf.SparseTensor`, `tf.RaggedTensor`, `tf.data.Dataset`, and
+`tf.StructuredTensor`.
+
+At a high level, types supported by this interface can be divided into two broad
+categories:
+
+* **_General data structures_**.
+  These types are handled by "generic" APIs whose behavior does not depend on
+  the value of each object (such as `tf.function`, `SavedModel`, and
+  `tf.while_loop`).
+
+* **_Tensor-like types_**, which specialize or extend tf.Tensor.
+  Values of these types have a `rank`, a `shape`, and usually a `dtype`.  In
+  addition to the "generic" APIs, these types can be handled by Tensor-specific
+  APIs (such as `tf.stack`, `tf.add`, and `tf.reduce_mean`).
+
+Examples of user-defined types that could defined or extended with this protocol
+include:
+
+**General data structures:**
+
+* `tfp.Distribution`: Encodes a statistical distribution.
+* `TensorDigraph`: Encodes the set of nodes and edges in a directed graph.
+* `DimensionAlignment`: Encodes a correspondence between two related dimensions
+  (e.g., between a `word` dimension and a `speaker` dimension).
+
+**Tensor-like types:**
+
+* `CSRSparseTensor`: A sparsely-encoded tensor that uses the Compressed Sparse
+  Row encoding.
+* `MaskedTensor`: Pairs a Tensor with a corresponding boolean mask, indicating
+  which values are valid, and automatically updates the mask as appropriate
+  when used with TensorFlow ops (such as `tf.add` or `tf.reduce_sum`).
+* `LabeledTensor`: Pairs a Tensor with a list of axis names, which can be used
+  for error detection and reporting, for specifying axes by name, and for
+  broadcasting.
+
+## User Benefit
+
+### Object-Oriented TensorFlow
+
+This proposal brings the benefits of Object-Oriented Programming to TensorFlow
+users, allowing them to define modular encapsulated data structures that
+interoperate with TensorFlow APIs.  This allows TensorFlow models to be defined
+at a higher level of abstraction.
+
+### Development outside the main TensorFlow repository
+
+Prior to this proposal, the only way to develop such data structures (e.g.
+`SparseTensor`) was to develop them inside the main TensorFlow code base.  This
+introduced significant barriers to rapid development, including slow release
+cycles, strong backwards compatibility constraints, and centralized API
+approval.  By allowing such data structures to be developed outside the main
+TensorFlow code base, we hope to make it much easier to experiment with new
+types and designs.  If general-purpose types are developed that become
+sufficiently mature, we may consider bringing them into the main TensorFlow code
+base.
+
```

Comment:
Have you seen this? 

https://github.com/tensorflow/probability/blob/b3ea633bdf9b2dbbbe4fcb82b0dd805cb9f284f8/tensorflow_probability/python/util/deferred_tensor.py#L81

https://github.com/tensorflow/probability/blob/b3ea633bdf9b2dbbbe4fcb82b0dd805cb9f284f8/tensorflow_probability/python/util/deferred_tensor.py#L117

Looks like an already existing example of what this RFC is addressing, would be interesting to see the details of this usecase.